cpp/include/cuml/experimental/fil/detail/device_initialization/gpu.cuh

```diff
@@ -0,0 +1,358 @@
+      cudaFuncAttributeMaxDynamicSharedMemorySize,
+      max_shared_mem_per_block
+    )
+  );
+  raft_proto::cuda_check(
+    cudaFuncSetAttribute(
+      infer_kernel<true, 1, forest_t, typename forest_t::io_type*, std::nullptr_t>,
+      cudaFuncAttributeMaxDynamicSharedMemorySize,
+      max_shared_mem_per_block
+    )
+  );
+  raft_proto::cuda_check(
+    cudaFuncSetAttribute(
+      infer_kernel<true, 2, forest_t, typename forest_t::io_type*, std::nullptr_t>,
+      cudaFuncAttributeMaxDynamicSharedMemorySize,
+      max_shared_mem_per_block
+    )
+  );
+  raft_proto::cuda_check(
+    cudaFuncSetAttribute(
+      infer_kernel<true, 4, forest_t, typename forest_t::io_type*, std::nullptr_t>,
+      cudaFuncAttributeMaxDynamicSharedMemorySize,
+      max_shared_mem_per_block
+    )
+  );
+  raft_proto::cuda_check(
+    cudaFuncSetAttribute(
+      infer_kernel<true, 8, forest_t, typename forest_t::io_type*, std::nullptr_t>,
+      cudaFuncAttributeMaxDynamicSharedMemorySize,
+      max_shared_mem_per_block
+    )
+  );
+  raft_proto::cuda_check(
+    cudaFuncSetAttribute(
+      infer_kernel<true, 16, forest_t, typename forest_t::io_type*, std::nullptr_t>,
+      cudaFuncAttributeMaxDynamicSharedMemorySize,
+      max_shared_mem_per_block
+    )
+  );
+  raft_proto::cuda_check(
+    cudaFuncSetAttribute(
+      infer_kernel<true, 32, forest_t, typename forest_t::io_type*, std::nullptr_t>,
+      cudaFuncAttributeMaxDynamicSharedMemorySize,
+      max_shared_mem_per_block
+    )
+  );
+  raft_proto::cuda_check(
+    cudaFuncSetAttribute(
+      infer_kernel<true, 1, forest_t, typename forest_t::io_type*, typename forest_t::node_type::index_type*>,
+      cudaFuncAttributeMaxDynamicSharedMemorySize,
+      max_shared_mem_per_block
+    )
+  );
+  raft_proto::cuda_check(
+    cudaFuncSetAttribute(
+      infer_kernel<true, 2, forest_t, typename forest_t::io_type*, typename forest_t::node_type::index_type*>,
+      cudaFuncAttributeMaxDynamicSharedMemorySize,
+      max_shared_mem_per_block
+    )
+  );
+  raft_proto::cuda_check(
+    cudaFuncSetAttribute(
+      infer_kernel<true, 4, forest_t, typename forest_t::io_type*, typename forest_t::node_type::index_type*>,
+      cudaFuncAttributeMaxDynamicSharedMemorySize,
+      max_shared_mem_per_block
+    )
+  );
+  raft_proto::cuda_check(
+    cudaFuncSetAttribute(
+      infer_kernel<true, 8, forest_t, typename forest_t::io_type*, typename forest_t::node_type::index_type*>,
+      cudaFuncAttributeMaxDynamicSharedMemorySize,
+      max_shared_mem_per_block
+    )
+  );
+  raft_proto::cuda_check(
+    cudaFuncSetAttribute(
+      infer_kernel<true, 16, forest_t, typename forest_t::io_type*, typename forest_t::node_type::index_type*>,
+      cudaFuncAttributeMaxDynamicSharedMemorySize,
+      max_shared_mem_per_block
+    )
+  );
+  raft_proto::cuda_check(
+    cudaFuncSetAttribute(
+      infer_kernel<true, 32, forest_t, typename forest_t::io_type*, typename forest_t::node_type::index_type*>,
```

Comment:
How do we ensure that we didn't miss any instantiation of `infer_kernel` in the device initialization stage? Listing all instantiations manually seems quite error prone.

Reply:
That's why I (reluctantly) used a macro for the required explicit instantiations. We still need to make sure that we're instantiating for every index of the specialization variants, but if there's a problem (e.g. we've added new variants), it should be immediately obvious that missing indexes in those macro invocations are the cause.